pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/models.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import annotations
+
+import platform
+import re
+from dataclasses import dataclass
+from enum import Enum
+from pathlib import Path
+from typing import Any, Final, TypedDict
+
+
+class NRError(Exception):
+    "nixos-rebuild general error."
+
+    def __init__(self, message: str):
+        self.message = message
+
+    def __str__(self) -> str:
+        return f"error: {self.message}"
+
+
+class Action(Enum):
+    SWITCH = "switch"
+    BOOT = "boot"
+    TEST = "test"
+    BUILD = "build"
+    EDIT = "edit"
+    REPL = "repl"
+    DRY_BUILD = "dry-build"
+    DRY_RUN = "dry-run"
+    DRY_ACTIVATE = "dry-activate"
+    BUILD_VM = "build-vm"
+    BUILD_VM_WITH_BOOTLOADER = "build-vm-with-bootloader"
+    LIST_GENERATIONS = "list-generations"
+
+    def __str__(self) -> str:
+        return self.value
+
+    @staticmethod
+    def values() -> list[str]:
+        return [a.value for a in Action]
+
+
+@dataclass(frozen=True)
+class Flake:
+    path: Path
+    attr: str
+    _re: Final[re.Pattern[str]] = re.compile(r"^(?P<path>[^\#]*)\#?(?P<attr>[^\#\"]*)$")
+
+    def __str__(self) -> str:
+        return f"{self.path}#{self.attr}"
+
+    @classmethod
+    def parse(cls, flake_str: str) -> Flake:
+        m = cls._re.match(flake_str)
+        assert m is not None, f"got no matches for {flake_str}"
+        attr = m.group("attr")
+        if not attr:
+            hostname = platform.node() or "default"
```

Comment:
nit

Python's documentation says this:

> Returns the computer’s network name (may not be fully qualified!). An empty string is returned if the value cannot be determined.

I think we should safeguard against an empty value at least.

We should also consider moving this out of the model, as it makes this model impure. I personally would expect models/data classes to behave the same no matter the environment.

Reply:
> I think we should safeguard against an empty value at least.

The `or` should safeguard against empty value because empty strings in Python evaluates to false:

```
>>> print("" or "something")
something
```

Unless you meant to validate and return an error? Not sure if it would make much sense in this case.

> We should also consider moving this out of the model, as it makes this model impure. I personally would expect models/data classes to behave the same no matter the environment.

I am also conflicted by this, but this was the easiest way I found to keep compatibility with the logic from old `nixos-rebuild`.

I think it is fine, at least for now, especially because the model itself is still pure, this is just an alternative way to create an instance that is smarter about the logic.